### PR TITLE
Pokemon Summary Screen documentation

### PR DESCRIPTION
**Pull request by ravepossum**
## Pokemon Summary Screen documentation

This documents a lot of the summary screen and sets up the folder structure for the summary screen application. I'm using the word "application" here as laid out in #285.

I've named and begun to document the following files:
- `applications/pokemon_summary_screen/main.c` - the bulk of the application and the main state machine
- `applications/pokemon_summary_screen/sprite.c` - mostly sprite-related helper functionality
- `applications/pokemon_summary_screen/window.c` - mostly window-related helper functionality
- `applications/pokemon_summary_screen/subscreen.c` - mostly subscreen-related helper functionality
- `src/ribbon.c` - general ribbon functionality that is used in the summary screen

(there's at least one more file that will end up with the rest of the summary screen code, a primarily 3D graphics helper file `unk_020916B4.c`)

I'm going ahead and submitting this because the changes are growing a little unwieldy, but the documentation is very much a work-in-progress. There's still lots to finish like:
- Reviewing and refining the names of the main struct members
- Labeling many many magic numbers including windows, sprites, and message entry IDs
- Documenting the 3D graphics helper file
- Documenting the rest of the functions in each file

## Multi-file application organization

With this we're also proposing the following folder structure for multi-file applications:
```
src/
---applications/
--------pokemon_summary_screen/
------------main.c
------------sprite.c
------------window.c
--------options_menu.c
```

Multi-file applications would have their own directories, while self-contained applications would remain in the `src/applications` directory (and similar for the relevant header files). The main application that runs would go into a file called `main.c`. The naming patterns would *not* include the application name, as it would be included in the file path, e.g. `applications/pokemon_summary_screen/main.c`.

### CODE REVIEW

File: src/applications/pokemon_summary_screen/main.c [hunk 1]
```diff
@@ -0,0 +1,2174 @@
...
+
+    return TRUE;
+}
+
+static void PokemonSummaryScreenVBlank(void *data)
+{
+    PokemonSummaryScreen *summaryScreen = data;
+
+    Bg_RunScheduledUpdates(summaryScreen->bgConfig);
+    sub_02008A94(summaryScreen->monSprite.spriteManager);
```
Comment (lhearachel) [inline src/applications/pokemon_summary_screen/main.c:308]: This is probably similar to `AnimateSprites` in Emerald? Not that you need to name it at present.

File: src/applications/pokemon_summary_screen/main.c [hunk 2]
```diff
@@ -0,0 +1,2174 @@
...
+    Graphics_LoadTilesToBgLayerFromOpenNARC(narc, 0, summaryScreen->bgConfig, BG_LAYER_MAIN_3, 0, 0, FALSE, HEAP_ID_POKEMON_SUMMARY_SCREEN);
+    Graphics_LoadTilemapToBgLayerFromOpenNARC(narc, 3, summaryScreen->bgConfig, BG_LAYER_MAIN_3, 0, 0, FALSE, HEAP_ID_POKEMON_SUMMARY_SCREEN);
+    Graphics_LoadPaletteFromOpenNARC(narc, 1, PAL_LOAD_MAIN_BG, 0, 0, HEAP_ID_POKEMON_SUMMARY_SCREEN);
+
+    Graphics_LoadTilemapToBgLayerFromOpenNARC(narc, 11, summaryScreen->bgConfig, BG_LAYER_MAIN_2, 0, 0, FALSE, HEAP_ID_POKEMON_SUMMARY_SCREEN);
+
+    Graphics_LoadTilesToBgLayerFromOpenNARC(narc, 2, summaryScreen->bgConfig, BG_LAYER_SUB_0, 0, 0, FALSE, HEAP_ID_POKEMON_SUMMARY_SCREEN);
+    Graphics_LoadTilesToBgLayerFromOpenNARC(narc, 16, summaryScreen->bgConfig, BG_LAYER_SUB_1, 0, 0, FALSE, HEAP_ID_POKEMON_SUMMARY_SCREEN);
+    Graphics_LoadTilemapToBgLayerFromOpenNARC(narc, 15, summaryScreen->bgConfig, BG_LAYER_SUB_1, 0, 0, FALSE, HEAP_ID_POKEMON_SUMMARY_SCREEN);
+    Graphics_LoadPaletteFromOpenNARC(narc, 14, PAL_LOAD_SUB_BG, 0, 0, HEAP_ID_POKEMON_SUMMARY_SCREEN);
```
Comment (lhearachel) [inline src/applications/pokemon_summary_screen/main.c:467]: Are you planning to unpack the assets as part of your next PR? I want to make sure that these aren't left as magic numbers forever, but the `naix` for them obviously doesn't exist yet.

File: src/applications/pokemon_summary_screen/main.c [hunk 3]
```diff
@@ -0,0 +1,2174 @@
...
+        return PSS_STATE_HANDLE_INPUT;
+    }
+
+    if (gCoreSys.pressedKeys & PAD_BUTTON_B) {
+        Sound_PlayEffect(SEQ_SE_DP_DECIDE);
+        summaryScreen->data->returnMode = 1;
+        return PSS_STATE_TRANSITION_OUT;
+    }
+
+    if (gCoreSys.pressedKeys & PAD_BUTTON_A) {
```
Comment (lhearachel) [inline src/applications/pokemon_summary_screen/main.c:558]: `JOY_NEW(PAD_BUTTON_A)` (and similar elsewhere; these are in `core_sys.h`)

File: src/applications/pokemon_summary_screen/main.c [hunk 4]
```diff
@@ -0,0 +1,2174 @@
...
+        if (summaryScreen->data->mode == PSS_MODE_POFFIN && summaryScreen->page == PSS_MODE_CONDITION) {
+            Sound_PlayEffect(SEQ_SE_DP_DECIDE);
+            return TryFeedPoffin(summaryScreen);
+        }
+
+        if (summaryScreen->page == PSS_PAGE_BATTLE_MOVES) {
+            Sound_PlayEffect(SEQ_SE_DP_SYU01);
+            summaryScreen->subscreen = 0;
```
Comment (lhearachel) [inline src/applications/pokemon_summary_screen/main.c:566]: What does this value do? Is it related to the subscreen mode constants defined in the header?

File: src/applications/pokemon_summary_screen/main.c [hunk 5]
```diff
@@ -0,0 +1,2174 @@
...
+            summaryScreen->subscreen |= 16;
+        }
+
+        summaryScreen->monData.sheen = (u8)Pokemon_GetValue(mon, MON_DATA_SHEEN, NULL);
+
+        if (summaryScreen->data->dataType == PSS_DATA_BOX_MON) {
+            Heap_FreeToHeap(mon);
+        }
+
+        Font_LoadScreenIndicatorsPalette(0, 14 * 32, HEAP_ID_POKEMON_SUMMARY_SCREEN);
```
Comment (lhearachel) [inline src/applications/pokemon_summary_screen/main.c:926]: `palette.h` has this macro for semantically declaring palette slots:

```c
#define PLTT_OFFSET(i)     ((i) * PALETTE_SIZE_BYTES)
```

File: src/applications/pokemon_summary_screen/main.c [hunk 6]
```diff
@@ -0,0 +1,2174 @@
...
+
+    Pokemon_FromBoxPokemon(boxMon, mon);
+    SetMonDataFromMon(summaryScreen, mon, monData);
+    Heap_FreeToHeap(mon);
+}
+
+static void SetMonDataFromMon(PokemonSummaryScreen *summaryScreen, Pokemon *mon, PokemonSummaryMonData *monData)
+{
+    BOOL reencrypt = Pokemon_EnterDecryptionContext(mon);
+    monData->species = (u16)Pokemon_GetValue(mon, MON_DATA_SPECIES, NULL);
```
Comment (lhearachel) [inline src/applications/pokemon_summary_screen/main.c:1005]: Does this need an explicit cast?

File: src/applications/pokemon_summary_screen/main.c [hunk 7]
```diff
@@ -0,0 +1,2174 @@
...
+    if (Pokemon_IsShiny(mon) == TRUE) {
+        monData->isShiny = TRUE;
+    } else {
+        monData->isShiny = FALSE;
+    }
+
+    monData->ribbons[0] = 0;
+    monData->ribbons[1] = 0;
+    monData->ribbons[2] = 0;
+    monData->ribbons[3] = 0;
```
Comment (lhearachel) [inline src/applications/pokemon_summary_screen/main.c:1104]: Does this work as a `memset` or a full array initializer?

```c
monData->ribbons = { 0 };
```

File: src/applications/pokemon_summary_screen/main.c [hunk 8]
```diff
@@ -0,0 +1,2174 @@
...
+    summaryScreen->ribbonMax = 0;
+
+    for (i = 0; i < 80; i++) {
```
Comment (lhearachel) [inline src/applications/pokemon_summary_screen/main.c:1107]: `i < RIBBON_MAX`

File: src/applications/pokemon_summary_screen/main.c [hunk 9]
```diff
@@ -0,0 +1,2174 @@
...
+    case BARCOLOR_EMPTY:
+        v0 = 0xc0 | 0xa000;
+        break;
+    case BARCOLOR_YELLOW:
+        v0 = 0xe0 | 0xa000;
+        break;
+    case BARCOLOR_RED:
+        v0 = 0x100 | 0xa000;
+        break;
+    }
```
Comment (lhearachel) [inline src/applications/pokemon_summary_screen/main.c:1355]: This should just be the palette for a tile, right? `0xA000` for the palette (slot 10) and one of `0x0100`, `0x00E0`, or `0x00C0` for the actual tile number. Looking ahead, it seems like `v0` should thus be the base tile along with its palette (which is what that `17` in `Bg_FillTilemapRect` is for; see the constant definition in `bg_window.h`).

```c
#define TILEMAP_FILL_VAL_INCLUDES_PALETTE 17
```

File: src/applications/pokemon_summary_screen/main.c [hunk 10]
```diff
@@ -0,0 +1,2174 @@
...
+void *PokemonSummaryScreen_MonData(PokemonSummaryScreen *summaryScreen)
+{
+    switch (summaryScreen->data->dataType) {
+    case PSS_DATA_MON:
+        return (void *)((u32)summaryScreen->data->monData + (Pokemon_GetStructSize() * summaryScreen->data->pos));
+    case PSS_DATA_PARTY_MON:
+        return (void *)Party_GetPokemonBySlotIndex(summaryScreen->data->monData, summaryScreen->data->pos);
+    case PSS_DATA_BOX_MON:
+        return (void *)((u32)summaryScreen->data->monData + (BoxPokemon_GetStructSize() * summaryScreen->data->pos));
+    }
```
Comment (lhearachel) [inline src/applications/pokemon_summary_screen/main.c:1572]: Is the explicit `void*` cast necessary?

**Review by lhearachel — COMMENTED** [review]
Great work so far. :pray: I tried to dodge commenting on any functions that were unnamed, as I interpreted that as you not having given them a thorough look as of yet.

General comment: does `screen` add anything as terminology? Would it be better to call this `pokemon_summary`?

File: src/applications/pokemon_summary_screen/subscreen.c [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+        if (summaryScreen->data->showContest != FALSE) {
+            summaryScreen->subscreenType = PSS_SUBSCREEN_TYPE_NORMAL;
+        } else {
+            summaryScreen->subscreenType = PSS_SUBSCREEN_TYPE_NO_CONTEST_INFO;
+            Bg_ScheduleScroll(summaryScreen->bgConfig, BG_LAYER_SUB_0, BG_OFFSET_UPDATE_SET_X, 4);
+        }
+        break;
+    case PSS_MODE_SELECT_MOVE:
+    case PSS_MODE_POFFIN:
+    default:
```
[the remark this answers is not in the document]
Comment (ravepossum) [reply src/applications/pokemon_summary_screen/subscreen.c:106]: Surprisingly, yep.

File: src/applications/pokemon_summary_screen/subscreen.c [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+    u8 button = 0;
+
+    while (TRUE) {
+        if (buttonList[button].page == PSS_PAGE_NONE) {
+            break;
+        }
+
+        UpdateSubscreenButtonTilemap(summaryScreen, &buttonList[button], 0);
+        button++;
+    }
```
[the remark this answers is not in the document]
Comment (ravepossum) [reply src/applications/pokemon_summary_screen/subscreen.c:143]: Can't get it to match with a `for` loop.

File: src/applications/pokemon_summary_screen/subscreen.c [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+        return button;
+    }
+
+    return PSS_SUBSCREEN_BUTTON_NONE;
+}
+
+static int CheckSubscreenRectanglePress(PokemonSummaryScreen *summaryScreen)
+{
+    if (summaryScreen->subscreenType == PSS_SUBSCREEN_TYPE_NO_BUTTONS) {
+        return 0xffffffff;
```
[the remark this answers is not in the document]
Comment (ravepossum) [reply src/applications/pokemon_summary_screen/subscreen.c:162]: Not in the repo yet, but that would definitely make sense to create. It's used in several places in `touch_screen.c` as well.

File: src/applications/pokemon_summary_screen/main.c (hunk 4 above)
Comment (ravepossum) [reply src/applications/pokemon_summary_screen/main.c:566]: I don't think it actually has anything to do with the actual subscreen, it seems like it's used mostly as a sort of state indicator for some of the information that pops up on top of the rest of the screen (like the move info panel). It definitely needs a rename.

File: src/applications/pokemon_summary_screen/main.c (hunk 2 above)
Comment (ravepossum) [reply src/applications/pokemon_summary_screen/main.c:467]: No idea how to do that tbh, but I'll look into that!

**Comment by ravepossum** [comment]
Re: Pokemon Summary *Screen* vs Pokemon Summary

My initial thinking there was that it was a clear designator that it specifically refers to the whole UI/application and not summary information more generally (like the substruct currently called `PokemonSummary` that probably needs a rename). So just generally making it a bit less ambiguous.

I'm definitely open to chopping off "screen" if we don't think it contributes anything of value for the verbosity it adds.

**Comment by lhearachel** [comment]
> I'm definitely open to chopping off "screen" if we don't think it contributes anything of value for the verbosity it adds.

I think this is a fine explanation. `options_menu` has the same concept, where it's tied to another concept in the core game systems, and instead is just an app for modifying that data.

File: src/applications/pokemon_summary_screen/subscreen.c [hunk 14]
```diff
@@ -24,30 +24,30 @@ static void sub_0223D0BC(PokemonSummaryScreen *summaryScreen);
...
-    { PSS_PAGE_NONE, 0xFF, 0xFF, 0xFF }
+    { PSS_PAGE_INFO, 1, 1, 4 },
+    { PSS_PAGE_MEMO, 1, 2, 10 },
+    { PSS_PAGE_SKILLS, 2, 5, 15 },
+    { PSS_PAGE_BATTLE_MOVES, 3, 10, 18 },
+    { PSS_PAGE_CONDITION, 2, 17, 18 },
+    { PSS_PAGE_CONTEST_MOVES, 3, 22, 15 },
+    { PSS_PAGE_RIBBONS, 4, 25, 10 },
+    { PSS_PAGE_EXIT, 4, 26, 4 },
+    { PSS_PAGE_NONE, 255, 255, 255 }
```
Comment (lhearachel) [inline src/applications/pokemon_summary_screen/subscreen.c:35]: I would personally use `-1` for these max-value entries to more clearly signify that the value should not be trusted or used.

File: include/touch_screen.h [hunk 15]
```diff
@@ -2,6 +2,7 @@
 #define POKEPLATINUM_TOUCH_SCREEN_H
 
 #define TOUCHSCREEN_USE_CIRCLE 0xfe
+#define TOUCHSCREEN_INPUT_NONE 0xFFFFFFFF
```
Comment (lhearachel) [inline include/touch_screen.h:5]: As with another comment, probably just `-1`.

**Review by lhearachel — APPROVED** [review]
This looks great! Nice work on getting this moving. I just have a short nitpick about using `-1` for constants meant to signify the concept of `NONE` or `INVALID`.

**Comment by ravepossum** [comment]
Merge conflicts resolved.

**Pull request merged**